Discussion: New Configuration Design Spec (#12)
adobe/helix-shared: CONFIGURATION.md

- the ability to quickly generate testable strains for code or content branches that are under development
- the ability to quickly and conveniently switch between strains in simulator and production
- the assurance that none of the above will break a published site
- the ability to do all of the above in a fully automated fashion on a CI system (e.g. generate and deploy a new strain for a new branch)

Site administration needs also include:
- the ability to define content mount points
- the ability to pass through traffic to another CMS (proxy strains)

Furthermore, we should establish:
- automated validation of configuration files that is independent from loading them
- detailed and generated documentation for all configuration properties

## Proposal: We keep YAML as the configuration format

While I have a good collection of scars from dealing with YAML (especially in CircleCI), I think the format is moderately more readable than JSON and almost as widely accepted, which makes it worthwhile suffering through it instead of picking something that will feel uncommon or out of the time.

## Proposal: We introduce a JSON Schema for configuration files

> This way you can see errors in configuration files while editing them, and we have a formal way of describing and documenting the configuration format.

Reviewer: +1

note: we can "only" validate the resulting Object structure (i.e. the `toJson()`) of the yaml. which might give confusing errors. but having validation is better that nothing. the validation should/could happen in the during `HelixConfig.parse()`

Author: https://github.com/redhat-developer/vscode-yaml also supports Schema validation on the input.

## Proposal: Strains are the only top-level concept

This means that there are no concepts that are orthogonal to strains, but that every other configuration object can be assigned to one or more strains, and that every request can be assigned to one specific strain.

> 

Reviewer: what about the mount points? is it possible to have the yaml in a way to define the mountpoints ourside the strains and then "magically" reference them?

Author: Yes, with a `definition` container that contains random referenceable config objects.

      address: 192.168.0.1
      name: publish
      use_ssl: false
```

## Proposal: Configuration objects can be referenced using YAML references

In order to re-use a complex configuration object (like an origin, a mount mapping, or even a strain), use YAML references (`*publish`) or YAML extensions (`<< *basestrain`). This allows us to rely on the behavior of the YAML parser instead of using home-grown inheritance and referencing logic that is likely less robust and certainly less well documented.

A strain re-using the `*basestrain` above would look like this:

```yaml
  client:
    <<: *basestrain
    condition: req.http.host == "client.project-helix.io"
    content:
      repo: helix-cli
      ref: master
      owner: adobe
    directoryIndex: readme.html
```

A strain referencing the `*publish` origin would look like this:
```yaml
  proxy-detailed:
    sticky: true
    origin: *publish
```

## Proposal: The `default` strain serves all traffic in production when no other strain is selected, therefore a `default` strain is mandatory

We need one strain that is serving traffic that no other strain claims. This could be one by:

1. requiring the existence of a `default` strain

> 2. requiring the existence of one and only one strain that has the `default: true` property set

Reviewer: 3. have a toplevel configuration value: `defaultStrain` that points to the default strain.

Author: (3) would be even harder to validate.

Option (1) is easier to validate than option (2) and less error prone (every YAML editor will tell you if you have two keys called `default`), so I'd opt for (1)

`hlx publish` should fail when no `default` strain exists.

> `hlx up` can continue without a `default` strain, but should issue a warning

Reviewer: +1 for option 1, since it easy allows to use yaml references in the `default` strain, eg: `default: *strains.master`

## Proposal: There is no implicit inheritance from `default`

All inheritance becomes explicit, using the mechanism laid out in 
https://github.com/adobe/helix-shared/issues/11#issuecomment-443122026. This means the `default` strain is special only in regards to serving traffic, but does not alter the behavior of any other strain.

## Proposal: Strains can have a `mount` configuration that maps URL paths to content paths

Right now, strains can have only one `content` node. With support for `mount`, a strain can have either a `content` or a `mount` property. Each `mount` object is a map between the URL path (key) and a `content` property (either git URL or object).

```yaml
strains:
  default:
    mount:
      /: https://github.com/adobe/project-helix.io.git
      /pipeline: https://github.com/adobe/helix-pipeline.git
      /cli: https://github.com/adobe/helix-cli.git

> ```

Reviewer: why not have X strains for X mountpoints? otherwise you creating an implicit default for the `/` <-> `contentRepo` if there is no `mount` structure. with this approach, you cannot choose a different code repository for a mountpoint.

Author: We decided during the hackathon not to introduce mount points, so that it will be X strains for X mount points.

Reviewer 2: Just so I'm clear here, this means each strain has at most a single mount point, is that correct?

## Proposal: `mount` points can take Git URLs, `content` objects/references, Backend URLs, `origin` objects/references

> That way you can have a consistent traffic mapping without having to switch strains.

Reviewer: can you elaborate the `content` and `origin` references?

Author: `content` can be a Git URL (`string`) or an object with resolved properties. When it is an object, that object can be made referencable, and in YAML semantics pointing to the reference is as good as copying and pasting.

## Proposal: `hlx deploy` lists all strains that are affected by the deployment and suggests the creation of a new strain if none are affected

> 

Reviewer: this assumes that `hlx deploy` always acts on only 1 strain. which is either the name of the current branch, `dirty` or what is specified with `--strain`

Author: Yes, `hlx deploy` would generate a finite amount of deployment packages, and we can find out if there are referenced in the YAML.

> In order to increase the visibility of changes happening during deployment, `hlx deploy` will list all strain names that will be affected by the deployment.

Reviewer: who is this list determined? what selects a strain based on the current git-repo? note, that we can also have pure-config git-repos.

Author: If `hlx deploy` does not have access to `helix-config.yaml` we won't be able to build that list (this would be the case for a separate config repo). I would assume that in this case you already know what you are doing and don't need hand-holding in updating your config.

Reviewer: How is the list generated? how is it different from just 'all' strains? how does a strain select the actions?

> If no strains are affected, `hlx deploy` will print a new strain config to `stdout` that points to the new `code` location, copies all other values from `default` except for `url` or `condition`. The new strain will have an auto-generated, hard-to-guess name, so that it cannot unwittingly be accessed.

Reviewer: why is it important to have a random name? why not using the branch-name ?

Author: Two reasons:

1. to avoid conflicts
2. to prevent people from forging the `X-Strain` cookie and getting access to development- or staging-only strains.

For the most part, the default logic of the current `X-Strain` will be applied, so that the strain's directory index, static repo, etc. will be used. The only exception is the resolution of the OpenWhisk action to execute.
Here, the `-git--([\w]+)--` pattern will be replaced with the `X-Tag` value, effectively pinning the used action to the tagged deployment. 

As this will lead to a large number of temporary actions in OpenWhisk, we also introduce a `hlx undeploy` command that clears a temporary deployment made earlier. This can be run at the end of every CI job.

`hlx test` and `hlx perf` should use the new temporary strains when running in CI.

## Discussion: layout

During the 4/2018 hackathon we also discussed the layout of the project. it was decided that:

- templates and pre functions will be in `./src`
- client side javascript **must not** go into `./src`. 
- we will provide a react example, and put the sources in `./react`
- all static files will go to `htdocs`
- all additional openwhisk actions will go to `htdocs/cgi-bin`

## Discussion: the `code` property

> - the `code` property of a strain is either a giturl string or giturl object (but not an action name).

Reviewer: I'm still not sure how this works.

so assume a non-dirty checkout, `git@github.com/adobe/helix-io.git#master (923c1e6)`.
1. the default strain contains a code property: `code: "git@github.com/adobe/helix-io.git#master"`
2. `hlx deploy` would generate an action `/helix/923c1e6/html`
3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#923c1e6"`

assume a dirty checkout, `git@github.com/adobe/helix-io.git#master (a0c96a5)`
1. the default strain contains a code property: `code: "git@github.com/adobe/helix-io.git#master"`
2. `hlx deploy --dirty` would generate an action `/helix/a0c96a5-dirty/html`
3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#a0c96a50-dirty"`<sup>1</sup>

<sup>1</sup> not sure about the `-dirty` suffix in the ref here. but fastly wouldn't have a way to figure it our

A strain re-using the `*basestrain` above would look like this:

```yaml
  client:
    <<: *basestrain
    condition: req.http.host == "client.project-helix.io"
    content:
      repo: helix-cli
      ref: master
      owner: adobe
    directoryIndex: readme.html
```

A strain referencing the `*publish` origin would look like this:
```yaml
  proxy-detailed:
    sticky: true
    origin: *publish
```

## Proposal: The `default` strain serves all traffic in production when no other strain is selected, therefore a `default` strain is mandatory

We need one strain that is serving traffic that no other strain claims. This could be one by:

1. requiring the existence of a `default` strain
2. requiring the existence of one and only one strain that has the `default: true` property set

Option (1) is easier to validate than option (2) and less error prone (every YAML editor will tell you if you have two keys called `default`), so I'd opt for (1)

`hlx publish` should fail when no `default` strain exists.

> `hlx up` can continue without a `default` strain, but should issue a warning

Reviewer: I think the `default` strain should be mandatory by schema. If we only allow no config or a valid config, we can provide a built-in default config that is used for no-config.

Furthermore, we should establish:
- automated validation of configuration files that is independent from loading them
- detailed and generated documentation for all configuration properties

## Proposal: We keep YAML as the configuration format

While I have a good collection of scars from dealing with YAML (especially in CircleCI), I think the format is moderately more readable than JSON and almost as widely accepted, which makes it worthwhile suffering through it instead of picking something that will feel uncommon or out of the time.

## Proposal: We introduce a JSON Schema for configuration files

This way you can see errors in configuration files while editing them, and we have a formal way of describing and documenting the configuration format. 

> Notes:
> - The validation is done during `HelixConfig.parse()`. 
> - Using a IDE plugin, the yaml can also be validated during input.

## Proposal: Strains are the only top-level concept

This means that there are no concepts that are orthogonal to strains, but that every other configuration object can be assigned to one or more strains, and that every request can be assigned to one specific strain.

> There will be no reconciliation of strains and some other concept at runtime. Every .hlx/strains.json will contain a fully deterministic, fully resolved and denormalized representation of all strains.

Reviewer 2: ```suggestion
There will be no reconciliation of strains and some other concept at runtime. Every `.hlx/strains.json` will contain a fully deterministic, fully resolved and denormalized representation of all strains.
```

      address: 192.168.0.1
      name: publish
      use_ssl: false
```

## Proposal: Configuration objects can be referenced using YAML references

In order to re-use a complex configuration object (like an origin, a mount mapping, or even a strain), use YAML references (`*publish`) or YAML extensions (`<< *basestrain`). This allows us to rely on the behavior of the YAML parser instead of using home-grown inheritance and referencing logic that is likely less robust and certainly less well documented.

A strain re-using the `*basestrain` above would look like this:

```yaml
  client:
    <<: *basestrain
    condition: req.http.host == "client.project-helix.io"
    content:
      repo: helix-cli
      ref: master
      owner: adobe
    directoryIndex: readme.html
```

A strain referencing the `*publish` origin would look like this:
```yaml
  proxy-detailed:
    sticky: true
    origin: *publish
```

## Proposal: The `default` strain serves all traffic in production when no other strain is selected, therefore a `default` strain is mandatory

We need one strain that is serving traffic that no other strain claims. This could be one by:

1. requiring the existence of a `default` strain

> 2. requiring the existence of one and only one strain that has the `default: true` property set

Author: Excellent question. The right answer is probably that the default strain must not have a condition.

      /cli: https://github.com/adobe/helix-cli.git
```
</del>

## Proposal: `mount` points can take Git URLs, `content` objects/references, Backend URLs, `origin` objects/references

> Notes:
> We decided during the 4/2018 hackathon not to introduce mount points, so that it will be X strains for X mount points.

<del>
That way you can have a consistent traffic mapping without having to switch strains.
</del>

## Discussion: the `code` property

- the `code` property of a strain is either a giturl string or giturl object (but not an action name).

assume a non-dirty checkout, `git@github.com/adobe/helix-io.git#master (923c1e6)`.
1. the default strain contains a code property: `code: "git@github.com/adobe/helix-io.git#master"`
2. `hlx deploy` would generate an action `/helix/923c1e6/html`
3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#923c1e6"`

assume a dirty checkout, `git@github.com/adobe/helix-io.git#master (a0c96a5)`
1. the default strain contains a code property: `code: "git@github.com/adobe/helix-io.git#master"`
2. `hlx deploy --dirty` would generate an action `/helix/a0c96a5-dirty/html`

> 3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#a0c96a50-dirty"`<sup>1</sup>

Author: Makes sense to me.

> 3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#a0c96a50-dirty"`<sup>1</sup>

Reviewer: ... or is there another property `action` ?

> 3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#a0c96a50-dirty"`<sup>1</sup>

Author: The thing that will differ is the package name, as the action name is now simply the file name, sans extension. So for an ad-hoc deployment, I'd expect we create a package ad-hoc and you would point the strain to the package, not the action. Hence the name of the property should be `package.`

> 3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#a0c96a50-dirty"`<sup>1</sup>

Reviewer: this doesn't really add up. if `publish` should not rely on the `effective-helix-config.yaml` then the steps for a dirty deploy would be like:

```
$ hlx deploy
hlx will not deploy a working copy that has uncommitted changes. Re-run with flag --dirty to force.
$ hlx deploy --dirty
no matching strains found for ad-hoc deployment. use --add=adhoc to create one or add the following to your config:

  adhoc:
    code: https://github.com/project-helix.io.git#master
    package: 75f29aa936bfc2b84bde5ac0ee4afbf824b1391e-dirty
```

> 3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#a0c96a50-dirty"`<sup>1</sup>

Reviewer: but this would imply that `hlx publish` only uses the strains of the current repo and the one that have an explicit `package` property. @trieloff  correct?

> 3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#a0c96a50-dirty"`<sup>1</sup>

Reviewer: also, `hlx publish` now needs to know the wsk namespace to generate the action prefix

### Proposal: No implicit defaults

Therefore, a radical proposal: **we won't have any implicit defaults**. 
At all. Strains won't magically inherit configuration values from:
- the global properties in the `helix-config.yaml`
- the command line
- application (CLI) provided defaults
- runtime (Fastly) provided defaults

Unless you configure or reference a configuration value directly, it won't be effective. 
Unless all required configuration values are set, you won't be able to deploy or publish.

### Proposal: Generate a usable default configuration

The flipside of this is that in order to enable development without any upfront configuration, we need to generate a usable default configuration taking the current context into consideration.

A default configuration should be suggested when:
- running `hlx deploy`
- running `hlx publish`
- running `hlx perf`
- running `hlx demo`
- outside of a CI environment

A default configuration should be saved when:
- running `hlx * --save`
- running `hlx up`

The default configuration should have two strains:
- `default` pointing to the git remote

> - `dev` pointing to localhost

Reviewer: @trieloff  so by default, `hlx up` should select the `--domain localhost` ?

Author: Yes.

Here, the `-git--([\w]+)--` pattern will be replaced with the `X-Tag` value, effectively pinning the used action to the tagged deployment. 

As this will lead to a large number of temporary actions in OpenWhisk, we also introduce a `hlx undeploy` command that clears a temporary deployment made earlier. This can be run at the end of every CI job.

`hlx test` and `hlx perf` should use the new temporary strains when running in CI.

## Discussion: layout

During the 4/2018 hackathon we also discussed the layout of the project. it was decided that:

- templates and pre functions will be in `./src`
- client side javascript **must not** go into `./src`. 
- we will provide a react example, and put the sources in `./react`
- all static files will go to `htdocs`
- all additional openwhisk actions will go to `htdocs/cgi-bin`

## Discussion: action names

The deployment mechanism changes so that the **sha of the code repository** at deployment time will become the package name. the action name will simply be the script name without any extensions.

>  

Reviewer: ```suggestion

**Note** the `--prefix` option for `hlx deploy` will be removed.

```

for example: 

legacy name:
```
/helix/https---github-com-trieloff-helix-demo-git--master--html
```

new name:

```
/helix/dcfee99d45c6a2baed86c8e4921623c48b902522/html
```

> 

Reviewer: Question: for the managed case, we wanted to put all actions to the same namespace `helix-clients` and group them with packages by customer name. openwhisk can't have sub-folders in packages. so I would rather not use up the package for the SHA here, so that the packages still can be used for clients.

```suggest
/helix/dcfee99d45c6a2baed86c8e4921623c48b902522--html
```

The `--dirty` suffix will move to the package name.
 
for example:

```
/helix/dcfee99d45c6a2baed86c8e4921623c48b902522-dirty/html

> ```

Reviewer: ```suggestion
**Note** the `--prefix` option for `hlx deploy` will be removed.
```

Furthermore, we should establish:
- automated validation of configuration files that is independent from loading them
- detailed and generated documentation for all configuration properties

## Proposal: We keep YAML as the configuration format

While I have a good collection of scars from dealing with YAML (especially in CircleCI), I think the format is moderately more readable than JSON and almost as widely accepted, which makes it worthwhile suffering through it instead of picking something that will feel uncommon or out of the time.

## Proposal: We introduce a JSON Schema for configuration files

This way you can see errors in configuration files while editing them, and we have a formal way of describing and documenting the configuration format. 

> Notes:
> - The validation is done during `HelixConfig.parse()`. 
> - Using a IDE plugin, the yaml can also be validated during input.

## Proposal: Strains are the only top-level concept

This means that there are no concepts that are orthogonal to strains, but that every other configuration object can be assigned to one or more strains, and that every request can be assigned to one specific strain.

> 

Author: I think development should work similar to the temporary strains:

1. development maps the `localhost`-URL Space to a global URL-space, so that `https://localhost:3000/www.example.com/` is treated as equivalent (in strain resolution) to `https://www.example.com/`
2. for all strains where the `code` property matches (loosely) the current working copy, use the current working directory as code base
3. for all strains where the `content` property matches (loosely) the current working copy, use the current working directory as content base
4. for all others: ??? maybe proxy OpenWhisk/GitHub?

> 

Reviewer: > development maps the localhost-URL Space to a global URL-space, so that https://localhost:3000/www.example.com/ 

I would prefer to find another way that doesn't require link rewriting. i.e. keep the HTML output identical to production. 

maybe a `hlx up --domain www.example.com`

> 

Author: Good idea.

<del>
That way you can have a consistent traffic mapping without having to switch strains.
</del>

## Discussion: the `code` property

- the `code` property of a strain is either a giturl string or giturl object (but not an action name).

assume a non-dirty checkout, `git@github.com/adobe/helix-io.git#master (923c1e6)`.
1. the default strain contains a code property: `code: "git@github.com/adobe/helix-io.git#master"`
2. `hlx deploy` would generate an action `/helix/923c1e6/html`
3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#923c1e6"`

assume a dirty checkout, `git@github.com/adobe/helix-io.git#master (a0c96a5)`
1. the default strain contains a code property: `code: "git@github.com/adobe/helix-io.git#master"`
2. `hlx deploy --dirty` would generate an action `/helix/a0c96a5-dirty/html`
3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#a0c96a50-dirty"`<sup>1</sup>

<sup>1</sup> not sure about the `-dirty` suffix in the ref here. but fastly wouldn't have a way to figure it our

> ## Proposal: `hlx deploy` lists all strains that are affected by the deployment and suggests the creation of a new strain if none are affected

Reviewer 3: I agree with Alex. `deploy` should probably still be a separate command for users who want to verify stuff on I/O Runtime before publishing, but it could be run implicitly by `publish` (if not run before).

> ## Proposal: `hlx deploy` lists all strains that are affected by the deployment and suggests the creation of a new strain if none are affected

Author: I think the separation of commands (see `hlx package`) is a good idea, but we should also run stuff for the user, so that a `hlx publish` in a fresh project starts a chain of `build`, `package`, `deploy`.

> ## Proposal: `hlx deploy` lists all strains that are affected by the deployment and suggests the creation of a new strain if none are affected

Reviewer 4: I'd rather see that as "simple" vs "super advanced / specific usage": 90% of the users might not need / want to know about `build` `package` `deploy`...

> Notes:
> - `hlx deploy` would generate a finite amount of deployment packages, and we can find out if there are referenced in the YAML.

In order to increase the visibility of changes happening during deployment, `hlx deploy` will list all strain names that will be affected by the deployment.

`hlx deploy` gets the current git-remote `git remote get-url origin` as `$CURRENT_CODE_REPO` and checks all strains for a `code` property that matches the `$CURRENT_CODE_REPO`.

If no strains are affected, `hlx deploy` will print a new strain config to `stdout` that points to the new `code` location (`$CURRENT_CODE_REPO`), copies all other values from `default` except for `url` or `condition`. The new strain will have an auto-generated, hard-to-guess name, so that it cannot unwittingly be accessed.

When running `hlx deploy --add=foo` the new strain will be added to the configuration file automatically and `hlx deploy` will instead show instructions on accessing the strain.

When running `hlx deploy --add=default` the default strain will be created or updated.

> 

Author: It should always merge, i.e. update the `code` property, and keep everything else the way it was.

> Notes:
> - Q: why is it important to have a random name? why not using the branch-name ?
> - A: Two reasons:
>      - to avoid conflicts
>      - to prevent people from forging the X-Strain cookie and getting access to development- or staging-only strains.

A deployment that does not affect any strains will have a non-zero exit code, so that it can fail in CI.

> 

Reviewer: Question: how to deal with https vs ssh giturls? i.e. `git remote get-url origin` might return a ssh giturl, but the yaml config should only use https ones...

> - the assurance that none of the above will break a published site
> - the ability to do all of the above in a fully automated fashion on a CI system (e.g. generate and deploy a new strain for a new branch)

### Proposal: No implicit defaults

Therefore, a radical proposal: **we won't have any implicit defaults**. 
At all. Strains won't magically inherit configuration values from:
- the global properties in the `helix-config.yaml`
- the command line
- application (CLI) provided defaults
- runtime (Fastly) provided defaults

Unless you configure or reference a configuration value directly, it won't be effective. 
Unless all required configuration values are set, you won't be able to deploy or publish.

### Proposal: Generate a usable default configuration

The flipside of this is that in order to enable development without any upfront configuration, we need to generate a usable default configuration taking the current context into consideration.

A default configuration should be suggested when:
- running `hlx deploy`
- running `hlx publish`
- running `hlx perf`
- running `hlx demo`
- outside of a CI environment

> A default configuartion should be saved when:

Author: ```suggestion
A default configuration should be saved when:
```

> - running `hlx * --save`

Author: As it is the default, I'd say it just overwrites what has been there before (or even refuses to overwrite, so that `--save` only works when there is no `helix-config.yaml`.

Concurrent deployments from a CI environment pose a hard problem at the moment:

1. if the strain configuration cannot be modified by the deployment action, a new strain must be introduced for every new branch, prior to the creation of the branch, so that the branch can be deployed and tested
2. if the strain configuration can be modified by the deployment action, a new strain can be introduced, but as the `helix-config.yaml` is the single point of synchronization, deployment in one branch will automatically deactivate all strains that refer to concurrent deployments made in different branches. In active development, this will lead to race conditions and intermittent test failures because the strain you just deployed got "undeployed" by a build triggered in another branch.
3. even within one branch, making two committs in short succession will lead to the second commit tainting the deployment of the first build, with potentially misleading results. For instance a bug fix might get attributed to the first commit because this is where testing first succeeds, although the fix was deployed in the second commit

### Proposal: Temporary Strains

> We introduce a new option for `hlx deploy --temp <tag[]?>` that takes zero or more tags for temporary deployments. By default, it uses `CIRCLE_SHA1` as the tag, i.e. the SHA1 hash of the last commit of the current build. This ties the deployment to the individual commit and avoids accidential overwriting.

Reviewer: I think this section needs overhaul

Author: ```suggestion
For testing a deployment in a Continuous Integration environment, it can be useful to have strains that are not persisted in the `helix-config.yaml`, but can still be activated for testing.
```

> When running `hlx deploy --temp`, all strains that refer a code repository that differs from the current repository will be ignored to avoid unccessary deploys.

Author: ```suggestion
```

> All strains that refer to the current code repository will be deployed with a `code` prefix that ends with the value of `tag`, e.g. `/acapt/default/https---github-com-adobe-project-helix-io-git--master--` would become `/acapt/default/https---github-com-adobe-project-helix-io-git--35ef52a772a7656be3d31527f8e595e4e286a0d4--`. 

Author: ```suggestion
```

> Each strain that is thus modified will be transformed into a temporary strain where the name includes the old strain name and the new `tag`, e.g. `default/35ef52a772a7656be3d31527f8e595e4e286a0d4`, `client/35ef52a772a7656be3d31527f8e595e4e286a0d4`, `pipeline/35ef52a772a7656be3d31527f8e595e4e286a0d4`. 

Author: ```suggestion
```

> These modified strains will not be written into `.hlx/strains.json` or persisted in any other way.

Author: ```suggestion
```

> Instead, the strain resolution logic in VCL will be modified so that when an `X-Strain` cookie or header is present and the value of the header contains a `/`, both the `X-Strain` (name) and `X-Tag` (tag) will be parsed. 

Author: ```suggestion
To enable temporary strains, the strain resolution logic in VCL will be modified so that when an `X-Strain` cookie or header is present and the value of the header contains a `/`, both the `X-Strain` (name) and `X-Tag` (tag) will be parsed. 
```

For the most part, the default logic of the current `X-Strain` will be applied, so that the strain's directory index, static repo, etc. will be used. The only exception is the resolution of the OpenWhisk action to execute.

> Here, the `-git--([\w]+)--` pattern will be replaced with the `X-Tag` value, effectively pinning the used action to the tagged deployment. 

Author: ```suggestion
Here, the commit sha that makes up the OpenWhisk package name will be replaced with the `X-Tag` value, effectively pinning the used action to the tagged deployment. 
```

> As this will lead to a large number of temporary actions in OpenWhisk, we also introduce a `hlx undeploy` command that clears a temporary deployment made earlier. This can be run at the end of every CI job.

Author: ```suggestion
As this will lead to a large number of temporary actions in OpenWhisk, we also introduce a `hlx undeploy` command that clears a deployment (based on SHA) made earlier. This can be run at the end of every CI job.
```

## Proposal: We introduce a JSON Schema for configuration files

This way you can see errors in configuration files while editing them, and we have a formal way of describing and documenting the configuration format. 

> Notes:

> > - The validation is done during `HelixConfig.parse()`. 

Author: +1

      /pipeline: https://github.com/adobe/helix-pipeline.git
      /cli: https://github.com/adobe/helix-cli.git
```
</del>

## Proposal: `mount` points can take Git URLs, `content` objects/references, Backend URLs, `origin` objects/references

> Notes:
> We decided during the 4/2018 hackathon not to introduce mount points, so that it will be X strains for X mount points.

<del>
That way you can have a consistent traffic mapping without having to switch strains.
</del>

## Proposal: `hlx deploy` lists all strains that are affected by the deployment and suggests the creation of a new strain if none are affected

> Notes:

> > - `hlx deploy` always acts on only 1 strain. which is either the name of the current branch, dirty or what is specified with `--strain`

Author: This is not my understanding. `hlx deploy` creates multiple actions, and the full path of the action (package + name) is dependent on:
- file name, e.g. `html.htl` becomes `/*/html`
- current sha, e.g. `b894d631b417a1367a0c2be8ec8ed7688d176e70` becomes `/b894d631b417a1367a0c2be8ec8ed7688d176e70/*`
- `--dirty` flag, e.g. `/b894d631b417a1367a0c2be8ec8ed7688d176e70-dirty/*`

This deployment package can then be referenced from multiple strains, but at the deploy time, `hlx` does not need to have any knowledge of strains. In contrast, `hlx publish` needs to have knowledge of deployment packages, and we need to pass this knowledge on to the user or update the configuration, if needed.

Reviewer: yes. you are correct. deploy doesn't care about strains.

As this will lead to a large number of temporary actions in OpenWhisk, we also introduce a `hlx undeploy` command that clears a temporary deployment made earlier. This can be run at the end of every CI job.

`hlx test` and `hlx perf` should use the new temporary strains when running in CI.

## Discussion: layout

During the 4/2018 hackathon we also discussed the layout of the project. it was decided that:

- templates and pre functions will be in `./src`
- client side javascript **must not** go into `./src`. 
- we will provide a react example, and put the sources in `./react`
- all static files will go to `htdocs`
- all additional openwhisk actions will go to `htdocs/cgi-bin`

## Discussion: the `code` property

> - the `code` property points to the code repository

Author: + owner, and ref.

Reviewer: the `code` property of a strain is either a giturl string or giturl object (but not an action name)

Author: That's a better way to phrase it.
